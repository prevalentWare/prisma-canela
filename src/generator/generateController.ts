import type { ParsedModel } from '../parser/types';
import type { Context } from 'hono';
import { Prisma } from '@prisma/client'; // Import Prisma namespace
import { pascalCase } from '../utils/pascalCase';
// import { camelCase } from "../utils/camelCase"; // Not strictly needed here anymore
import type { ServiceFunctionNames, ZodSchemaDetails } from './types';

// Helper function to generate controller imports
export const generateControllerImports = (
  model: ParsedModel,
  zodSchemaInfo: ZodSchemaDetails
): string => {
  const modelNamePascal = pascalCase(model.name);

  // Remove Zod and schema imports, as they are not used in the controller
  const zodImports = '';

  // Add validator utility function to safely cast validation results
  const utilityFunctions = `
// Helper to safely cast validation results - this avoids TypeScript errors with c.req.valid()
const getValidData = (c: Context, type: 'json' | 'param') => {
  // @ts-ignore - The Hono validation is properly set up by middleware, but TypeScript doesn't recognize it
  return c.req.valid(type);
};`;

  // Import service functions using namespace
  const serviceImports = `import * as service from './service';`;

  return `
import type { Context } from 'hono';
import { Prisma } from '@prisma/client';
${zodImports}
${utilityFunctions}
${serviceImports}
`;
};

// Helper function to generate handler logic
export const generateHandler = (
  modelNamePascal: string,
  handlerName: string, // e.g., listUser, createUserById
  serviceFunctionName: string, // e.g., findManyUsers, findUserById
  successStatusCode: number = 200,
  requiresId: boolean = false,
  requiresBody: boolean = false,
  idType: string = 'string' // Default, only used if requiresId is true
): string => {
  // No need for complex Context typing - we'll use the utility function
  const contextType = 'Context';

  // Determine how to access validated data
  const paramValidation = requiresId
    ? `const { id } = getValidData(c, 'param');`
    : '';
  const jsonValidation = requiresBody
    ? `const data = getValidData(c, 'json');`
    : '';

  // Determine arguments for service call with context as first parameter
  const args = ['c']; // Always pass context first
  if (requiresId) args.push('id');
  if (requiresBody) args.push('data');
  const serviceCallArgs = args.join(', ');

  const logId = requiresId ? ' ${id}' : '';
  const isIdNotFoundCheck = requiresId && !handlerName.startsWith('list'); // Check for P2025 on ID routes

  // Specific check for findById returning null
  const getByIdNotFound = handlerName.startsWith('get')
    ? `\n    if (!item) {\n      return c.json({ error: '${modelNamePascal} not found' }, 404);\n    }`
    : '';

  // Generate a natural language comment for the handler
  let comment = '';
  if (handlerName.startsWith('list')) {
    comment = `List all ${modelNamePascal} records.`;
  } else if (handlerName.startsWith('create')) {
    comment = `Create a new ${modelNamePascal}.`;
  } else if (handlerName.startsWith('get')) {
    comment = `Get a ${modelNamePascal} by ID.`;
  } else if (handlerName.startsWith('update')) {
    comment = `Update a ${modelNamePascal} by ID.`;
  } else if (handlerName.startsWith('delete')) {
    comment = `Delete a ${modelNamePascal} by ID.`;
  } else {
    comment = `${handlerName} for ${modelNamePascal}.`;
  }

  // Use the input generic in the handler signature
  // Only include paramValidation and jsonValidation if they are non-empty
  const validationLines = [paramValidation, jsonValidation]
    .filter(Boolean)
    .join('\n  ');

  return `
/**
 * ${comment}
 */
export const ${handlerName} = async (c: ${contextType}) => {
  ${validationLines}
  try {
    const item = await service.${serviceFunctionName}(${serviceCallArgs}); // Pass context first, then other args
    ${getByIdNotFound}
    // Always return status code explicitly for better type matching with openapi()
    return c.json(item, ${successStatusCode});
  } catch (error: unknown) {${
    isIdNotFoundCheck
      ? `\n    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {\n      // Use the specific ID in the error message if available\n      const message = \`Error ${handlerName
          .replace(/([A-Z])/g, ' $1')
          .toLowerCase()}ing ${modelNamePascal}${logId}: Record not found\`;
      console.error(message, error);\n      return c.json({ error: '${modelNamePascal} not found' }, 404);\n    }`
      : ''
  }
    // Check for Prisma client not found in context
    if (error instanceof Error && error.message.includes('Prisma client not found in context')) {
      console.error('Prisma client not found in context. Make sure to use the prismaMiddleware.');
      return c.json({ error: 'Database connection error. Please try again later.' }, 500);
    }
    
    // Generic error handling for other cases
    const message = \`Error ${handlerName
      .replace(/([A-Z])/g, ' $1')
      .toLowerCase()}ing ${modelNamePascal}${logId}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to ${handlerName
      .split(/(?=[A-Z])/)[0]
      .toLowerCase()} ${modelNamePascal}\` }, 500);
  }
};
`;
};

/**
 * Generates the content for a controller file (controller.ts).
 */
export const generateControllerFileContent = (
  model: ParsedModel,
  zodSchemaInfo: ZodSchemaDetails,
  serviceNames: ServiceFunctionNames
): string => {
  const modelNamePascal = pascalCase(model.name);
  const idField = model.fields.find((f) => f.isId);
  const idType = idField?.type === 'number' ? 'number' : 'string';

  const imports = generateControllerImports(model, zodSchemaInfo);

  let handlers = '';

  // List
  handlers += generateHandler(
    modelNamePascal,
    `list${modelNamePascal}`,
    serviceNames.findMany,
    200,
    false, // requiresId
    false // requiresBody
  );

  // Create
  handlers += generateHandler(
    modelNamePascal,
    `create${modelNamePascal}`,
    serviceNames.create,
    201,
    false, // requiresId
    true // requiresBody
  );

  // ID-based routes
  if (idField) {
    // Get By ID
    handlers += generateHandler(
      modelNamePascal,
      `get${modelNamePascal}ById`,
      serviceNames.findById,
      200,
      true, // requiresId
      false, // requiresBody
      idType
    );
    // Update
    handlers += generateHandler(
      modelNamePascal,
      `update${modelNamePascal}`,
      serviceNames.update,
      200,
      true, // requiresId
      true, // requiresBody
      idType
    );
    // Delete
    handlers += generateHandler(
      modelNamePascal,
      `delete${modelNamePascal}`,
      serviceNames.delete,
      200,
      true, // requiresId
      false, // requiresBody
      idType
    );
  }

  return imports + handlers;
};
