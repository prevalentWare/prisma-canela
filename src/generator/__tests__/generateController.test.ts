import { describe, it, expect } from "vitest";
import { generateControllerFileContent } from "../generateController";
import type { ParsedModel } from "../../parser/types";
import type { ZodSchemaDetails, ServiceFunctionNames } from "../types";
import * as z from "zod"; // Need z for inferring types

// --- Mock Data ---

// Shared Zod Schema Info (example for a 'User' model)
const userZodSchemaInfo: ZodSchemaDetails = {
  imports: [
    "import { UserSchema, createUserSchema, updateUserSchema } from './schema';",
    "import { Role } from '@prisma/client';", // Example enum import
  ],
  modelSchemaName: "UserSchema",
  createSchemaName: "createUserSchema",
  updateSchemaName: "updateUserSchema",
};

// Shared Service Function Names (example for a 'User' model)
const userServiceNames: ServiceFunctionNames = {
  findMany: "findManyUsers",
  findById: "findUserById",
  create: "createUser",
  update: "updateUser",
  delete: "deleteUser",
};

// 1. Standard Model (User with string ID)
const userModel: ParsedModel = {
  name: "User",
  dbName: null,
  fields: [
    {
      name: "id",
      type: "string",
      kind: "scalar",
      isId: true,
      isRequired: true,
      isUnique: true,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "email",
      type: "string",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: true,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "name",
      type: "string",
      kind: "scalar",
      isId: false,
      isRequired: false,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "role",
      type: "enum",
      kind: "enum",
      enumName: "Role",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
  ],
};

// 2. Model with Numeric ID (Product)
const productModel: ParsedModel = {
  name: "Product",
  dbName: null,
  fields: [
    {
      name: "productId",
      type: "number",
      kind: "scalar",
      isId: true,
      isRequired: true,
      isUnique: true,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "name",
      type: "string",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "price",
      type: "number",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
  ],
};
const productZodSchemaInfo: ZodSchemaDetails = {
  imports: [],
  modelSchemaName: "ProductSchema",
  createSchemaName: "createProductSchema",
  updateSchemaName: "updateProductSchema",
};
const productServiceNames: ServiceFunctionNames = {
  findMany: "findManyProducts",
  findById: "findProductById",
  create: "createProduct",
  update: "updateProduct",
  delete: "deleteProduct",
};

// 3. Model without ID (LogEntry)
const logEntryModel: ParsedModel = {
  name: "LogEntry",
  dbName: null,
  fields: [
    {
      name: "message",
      type: "string",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "level",
      type: "string",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
    {
      name: "timestamp",
      type: "date",
      kind: "scalar",
      isId: false,
      isRequired: true,
      isUnique: false,
      isList: false,
      hasDefaultValue: false,
    },
  ],
};
const logEntryZodSchemaInfo: ZodSchemaDetails = {
  imports: [],
  modelSchemaName: "LogEntrySchema",
  createSchemaName: "createLogEntrySchema",
  updateSchemaName: "updateLogEntrySchema",
};
const logEntryServiceNames: ServiceFunctionNames = {
  findMany: "findManyLogEntries",
  findById: "findLogEntryById", // Still need this for the type even if not used
  create: "createLogEntry",
  update: "updateLogEntry", // Still need this for the type even if not used
  delete: "deleteLogEntry", // Still need this for the type even if not used
};

// --- Test Suite ---

describe("generateControllerFileContent", () => {
  it("should generate correct controller content for a standard model (User)", () => {
    const result = generateControllerFileContent(
      userModel,
      userZodSchemaInfo,
      userServiceNames
    );
    // --- Assertions ---
    // Check imports
    expect(result).toContain("import type { Context } from 'hono';");
    expect(result).toContain("import * as service from './service';"); // Check service import
    expect(result).toContain(
      "import type { createUserSchema, updateUserSchema } from './schema';"
    );
    expect(result).toContain(
      "type CreateInput = z.infer<typeof createUserSchema>;"
    );
    expect(result).toContain(
      "type UpdateInput = z.infer<typeof updateUserSchema>;"
    );
    expect(result).toContain("import { Prisma } from '@prisma/client';"); // Check Prisma import for errors

    // Check function definitions (Updated Signatures)
    expect(result).toContain("export const listUser = async (c: Context)");
    expect(result).toContain("export const createUser = async (c: Context)"); // Simplified check
    expect(result).toContain("export const getUserById = async (c: Context)"); // Simplified check
    expect(result).toContain("export const updateUser = async (c: Context)"); // Simplified check
    expect(result).toContain("export const deleteUser = async (c: Context)"); // Simplified check

    // Check service function calls
    expect(result).toContain("await service.findManyUsers(c);");
    expect(result).toContain("await service.createUser(c, data);");
    expect(result).toContain("await service.findUserById(c, id);");
    expect(result).toContain("await service.updateUser(c, id, data);");
    expect(result).toContain("await service.deleteUser(c, id);");

    // Check basic ID retrieval (string ID)
    expect(result).toContain("const { id } = getValidData(c, 'param');"); // Updated assertion for param destructuring

    // Check error handling for Prisma known request errors (e.g., P2025)
    expect(result).toContain(
      "if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025')"
    );
    expect(result).toContain(
      "return c.json({ error: 'User not found' }, 404);"
    );

    // Check generic error handling
    expect(result).toContain(
      "console.error(message, error); // Log the detailed error"
    );
    expect(result).toContain("return c.json({ error: `Failed to");

    expect(result).toMatchSnapshot(); // Keep snapshot check
  });

  it("should generate correct controller content for a model with numeric ID (Product)", () => {
    const result = generateControllerFileContent(
      productModel,
      productZodSchemaInfo,
      productServiceNames
    );
    // --- Assertions ---
    // Check imports (similar to User, but with Product schemas)
    expect(result).toContain("import type { Context } from 'hono';");
    expect(result).toContain("import * as service from './service';");
    expect(result).toContain(
      "import type { createProductSchema, updateProductSchema } from './schema';"
    );
    expect(result).toContain(
      "type CreateInput = z.infer<typeof createProductSchema>;"
    );
    expect(result).toContain(
      "type UpdateInput = z.infer<typeof updateProductSchema>;"
    );
    expect(result).toContain("import { Prisma } from '@prisma/client';");

    // Check function definitions (Updated Signatures)
    expect(result).toContain("export const listProduct = async (c: Context)");
    expect(result).toContain("export const createProduct = async (c: Context)"); // Simplified check
    expect(result).toContain(
      "export const getProductById = async (c: Context)"
    ); // Simplified check
    expect(result).toContain("export const updateProduct = async (c: Context)"); // Simplified check
    expect(result).toContain("export const deleteProduct = async (c: Context)"); // Simplified check

    // Check service function calls
    expect(result).toContain("await service.findManyProducts(c);");
    expect(result).toContain("await service.createProduct(c, data);");
    expect(result).toContain("await service.findProductById(c, id);");
    expect(result).toContain("await service.updateProduct(c, id, data);");
    expect(result).toContain("await service.deleteProduct(c, id);");

    // Check numeric ID retrieval - relaxed check
    expect(result).toContain("const { id } = getValidData(c, 'param');"); // Type is handled by route validation

    // Check error handling
    expect(result).toContain(
      "if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025')"
    );
    expect(result).toContain(
      "return c.json({ error: 'Product not found' }, 404);"
    );

    expect(result).toMatchSnapshot();
  });

  it("should generate correct controller content for a model without an ID (LogEntry)", () => {
    const result = generateControllerFileContent(
      logEntryModel,
      logEntryZodSchemaInfo,
      logEntryServiceNames
    );
    // --- Assertions ---
    // Check imports
    expect(result).toContain("import type { Context } from 'hono';");
    expect(result).toContain("import * as service from './service';");
    expect(result).toContain(
      "import type { createLogEntrySchema, updateLogEntrySchema } from './schema';" // Still expect update schema import even if not used in controller
    );
    expect(result).toContain(
      "type CreateInput = z.infer<typeof createLogEntrySchema>;"
    );
    // No UpdateInput type needed if no update handler

    // Check that only list and create functions are defined (Updated Signatures)
    expect(result).toContain("export const listLogEntry = async (c: Context)");
    expect(result).toContain(
      "export const createLogEntry = async (c: Context)"
    ); // Simplified check

    // Check service function calls
    expect(result).toContain("await service.findManyLogEntries(c);");
    expect(result).toContain("await service.createLogEntry(c, data);");

    // Check that ID-specific handlers are NOT present
    expect(result).not.toContain("export const getLogEntryById = async");
    expect(result).not.toContain("export const updateLogEntry = async");
    expect(result).not.toContain("export const deleteLogEntry = async");
    expect(result).not.toContain("findLogEntryById(id)"); // Check no service calls for ID ops
    expect(result).not.toContain("updateLogEntry(id, data)");
    expect(result).not.toContain("deleteLogEntry(id)");

    expect(result).toMatchSnapshot();
  });
});
