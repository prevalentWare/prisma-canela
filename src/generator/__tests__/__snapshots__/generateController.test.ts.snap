// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateControllerFileContent > should generate correct controller content for a model with numeric ID (Product) 1`] = `
"
import type { Context } from 'hono';
import { Prisma } from '@prisma/client';
import type { z } from 'zod';
import type { createProductSchema, updateProductSchema } from './schema';
type CreateInput = z.infer<typeof createProductSchema>;
type UpdateInput = z.infer<typeof updateProductSchema>;
import * as service from './service';

/**
 * Handles list product Product.
 */
export const listProduct = async (c: Context) => {
  
  
  try {
    const item = await service.findManyProducts(); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error list producting Product: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to list Product\` }, 500);
  }
};

/**
 * Handles create product Product.
 */
export const createProduct = async (c: Context) => {
  
  const data = c.req.valid('json') as CreateInput; // Assume json validation is set up in routes
  try {
    const item = await service.createProduct(data); // Use service namespace and correct args
    
    return c.json(item, 201);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error create producting Product: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to create Product\` }, 500);
  }
};

/**
 * Handles get product by id Product.
 */
export const getProductById = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  
  try {
    const item = await service.findProductById(id); // Use service namespace and correct args
    
    if (!item) {
      return c.json({ error: 'Product not found' }, 404);
    }
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error get product by iding Product \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'Product not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error get product by iding Product \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to get Product\` }, 500);
  }
};

/**
 * Handles update product Product.
 */
export const updateProduct = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  const data = c.req.valid('json') as UpdateInput; // Assume json validation is set up in routes
  try {
    const item = await service.updateProduct(id, data); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error update producting Product \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'Product not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error update producting Product \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to update Product\` }, 500);
  }
};

/**
 * Handles delete product Product.
 */
export const deleteProduct = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  
  try {
    const item = await service.deleteProduct(id); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error delete producting Product \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'Product not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error delete producting Product \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to delete Product\` }, 500);
  }
};
"
`;

exports[`generateControllerFileContent > should generate correct controller content for a model without an ID (LogEntry) 1`] = `
"
import type { Context } from 'hono';
import { Prisma } from '@prisma/client';
import type { z } from 'zod';
import type { createLogEntrySchema, updateLogEntrySchema } from './schema';
type CreateInput = z.infer<typeof createLogEntrySchema>;
type UpdateInput = z.infer<typeof updateLogEntrySchema>;
import * as service from './service';

/**
 * Handles list log entry LogEntry.
 */
export const listLogEntry = async (c: Context) => {
  
  
  try {
    const item = await service.findManyLogEntries(); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error list log entrying LogEntry: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to list LogEntry\` }, 500);
  }
};

/**
 * Handles create log entry LogEntry.
 */
export const createLogEntry = async (c: Context) => {
  
  const data = c.req.valid('json') as CreateInput; // Assume json validation is set up in routes
  try {
    const item = await service.createLogEntry(data); // Use service namespace and correct args
    
    return c.json(item, 201);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error create log entrying LogEntry: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to create LogEntry\` }, 500);
  }
};
"
`;

exports[`generateControllerFileContent > should generate correct controller content for a standard model (User) 1`] = `
"
import type { Context } from 'hono';
import { Prisma } from '@prisma/client';
import type { z } from 'zod';
import type { createUserSchema, updateUserSchema } from './schema';
type CreateInput = z.infer<typeof createUserSchema>;
type UpdateInput = z.infer<typeof updateUserSchema>;
import * as service from './service';

/**
 * Handles list user User.
 */
export const listUser = async (c: Context) => {
  
  
  try {
    const item = await service.findManyUsers(); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error list usering User: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to list User\` }, 500);
  }
};

/**
 * Handles create user User.
 */
export const createUser = async (c: Context) => {
  
  const data = c.req.valid('json') as CreateInput; // Assume json validation is set up in routes
  try {
    const item = await service.createUser(data); // Use service namespace and correct args
    
    return c.json(item, 201);
  } catch (error: unknown) {
    // Generic error handling for other cases
    const message = \`Error create usering User: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to create User\` }, 500);
  }
};

/**
 * Handles get user by id User.
 */
export const getUserById = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  
  try {
    const item = await service.findUserById(id); // Use service namespace and correct args
    
    if (!item) {
      return c.json({ error: 'User not found' }, 404);
    }
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error get user by iding User \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'User not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error get user by iding User \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to get User\` }, 500);
  }
};

/**
 * Handles update user User.
 */
export const updateUser = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  const data = c.req.valid('json') as UpdateInput; // Assume json validation is set up in routes
  try {
    const item = await service.updateUser(id, data); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error update usering User \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'User not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error update usering User \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to update User\` }, 500);
  }
};

/**
 * Handles delete user User.
 */
export const deleteUser = async (c: Context) => {
  const { id } = c.req.valid('param'); // Assume param validation is set up in routes
  
  try {
    const item = await service.deleteUser(id); // Use service namespace and correct args
    
    return c.json(item);
  } catch (error: unknown) {
    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
      // Use the specific ID in the error message if available
      const message = \`Error delete usering User \${id}: Record not found\`;
      console.error(message, error);
      return c.json({ error: 'User not found' }, 404);
    }
    // Generic error handling for other cases
    const message = \`Error delete usering User \${id}: \${error instanceof Error ? error.message : 'Unknown error'}\`;
    console.error(message, error); // Log the detailed error
    return c.json({ error: \`Failed to delete User\` }, 500);
  }
};
"
`;
